examples/apps/windows/ClientControl.xaml.cpp

```diff
@@ -0,0 +1,189 @@
+using namespace Windows::UI::Xaml::Input;
+using namespace Windows::UI::Xaml::Media;
+using namespace Windows::UI::Xaml::Navigation;
+
+// The User Control item template is documented at http://go.microsoft.com/fwlink/?LinkId=234236
+
+std::atomic<int> ClientControl::_clientPort{ TalkConsts::DEF_CLIENT_PORT_INIT };
+
+ClientControl::ClientControl()
+{
+    InitializeComponent();
+
+    ServerPort->Text = DEF_SERVER_PORT.ToString();
+    auto clientPort = _clientPort.load();
+    ClientPort->Text = clientPort.ToString();
+}
+
+void
+ClientControl::Init(
+    IAsyncThreadNotify^  notify,
+    IMainPageUIElements^ mainPageUIElements)
+{
+    _notify = std::move(notify);
+    _mainPageUIElements = std::move(mainPageUIElements);
+}
+
+void
+ClientControl::ProtocolChanged(
+    Protocol protocol)
+{
+    _protocol = protocol;
+}
+
+void
+ClientControl::Connect_Click(
+    Object^          sender,
+    RoutedEventArgs^ e)
+{
+    try
+    {
+        auto clientArgs = ref new ClientArgs();
+
+        auto serverIP = ServerIP->Text;
+        if (serverIP->IsEmpty()) Exception::CreateException(E_INVALIDARG, "No Server IP");
+
+        // check valid chars of Ipv6 Address
+        if (std::all_of(serverIP->Data(), serverIP->Data() + serverIP->Length(),
+            [](wchar_t c)
+        {
+            if (c < '0') return false;
+            else if (c <= '9') return true; // '0' ~ '9'
+            else if (c == ':') return true; // ':'
+            else if (c < 'a') return false;
+            else if (c <= 'z') return true; // 'a' ~ 'z'
+            else return false;
+        }))
+            Exception::CreateException(E_INVALIDARG, "Not a valid Server IPv6 address");
+
+        clientArgs->ServerHostName = ref new HostName(serverIP);
+
+        if (ServerPort->Text->IsEmpty()) Exception::CreateException(E_INVALIDARG, "No Server Port");
+        clientArgs->ServerPort = ServerPort->Text;
+
+        auto clientIP = ClientIP->Text;
+        if (clientIP->IsEmpty()) Exception::CreateException(E_INVALIDARG, "No Client IP");
+
+        // check valid chars of Ipv6 Address
+        if (std::all_of(clientIP->Data(), clientIP->Data() + clientIP->Length(),
+            [](wchar_t c)
+        {
+            if (c < '0') return false;
+            else if (c <= '9') return true; // '0' ~ '9'
+            else if (c == ':') return true; // ':'
+            else if (c < 'a') return false;
+            else if (c <= 'z') return true; // 'a' ~ 'z'
+            else return false;
+        }))
+            Exception::CreateException(E_INVALIDARG, "Not a valid client IPv6 address");
+
+
+        clientArgs->ClientHostName = ref new HostName(clientIP);
+
+        if (ClientPort->Text->IsEmpty()) Exception::CreateException(E_INVALIDARG, "No Client Port");
+        clientArgs->ClientPort = ClientPort->Text;
+
+        auto cleintContext = Factory::CreateClientContext(_notify, clientArgs, _protocol);
+        cleintContext->Connect_Click(sender, e);
+        {
```

Comment:
I used std::mutex with std:lock_guard protection in the beginning, I used code block to specify the scope the protection. That code block is left behind and useless after I switched to std::atomic. Remove it - commit 44b3298.